### PR TITLE
lib: nrf_modem: add modem trace flash backend

### PR DESCRIPTION
**Pull request by eivindj-nordic**
* Update the trace backend interface to support flash backends
* Add nrf_modem_trace_data_size
* Add nrf_modem_trace_read
* Add nrf_modem_trace_clear
* Add nrf_modem_trace_event_handler_set

* Add the modem trace flash backend, based on the modem trace flash sample.

Signed-off-by: Eivind Jølsgard <eivind.jolsgard@nordicsemi.no>

[sdk-zephyr#1051](https://github.com/nrfconnect/sdk-zephyr/pull/1051)

### CODE REVIEW

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+	trace_fcb.f_sector_cnt = f_sector_cnt;
+	trace_fcb.f_sectors = trace_flash_sectors;
+
+	LOG_DBG("Sectors: %d, first sector: %p, sector size: %d",
+		f_sector_cnt, trace_flash_sectors, trace_flash_sectors[0].fs_size);
+
+	err = fcb_init(FIXED_PARTITION_ID(MODEM_TRACE), &trace_fcb);
+	if (err) {
+		LOG_ERR("fcb_init error: %d", err);
+		goto out;
```
Comment (lemrey) [inline lib/nrf_modem_lib/trace_backends/flash/flash.c:218]: same here

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+	}
+
+	flash_dev = flash_area_get_device(modem_trace_area);
+	if (flash_dev == NULL) {
+		LOG_ERR("Failed to get flash device\n");
+		return -ENODEV;
+	}
+
+	/* After a cold boot the magic will contain random values. */
+	if (magic != magic_val) {
```
Comment (lemrey) [inline lib/nrf_modem_lib/trace_backends/flash/flash.c:187]: Not sure whether we need this.
There are three cases,
- nothing is in flash
- flash contains fcb data (modem traces)
- flash contains something other than traces (fcb not installed, pre-existing data?)

In the first two cases, we don't need to erase anything.
In the last case, what will happen during `fcb_init` ?
If that's going to return an error, can we simply erase then and retry?

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+#if IS_ENABLED(CONFIG_FCB_ALLOW_FIXED_ENDMARKER)
+	.f_flags = FCB_FLAGS_CRC_DISABLED,
+#endif
+};
+
+/* Store in __noinit RAM to perserve in warm boot. */
+static __noinit uint32_t magic;
+static __noinit size_t stream_buf_written;
+static __noinit size_t trace_bytes_written;
+static __noinit uint8_t stream_buf[BUF_SIZE];
```
Comment (lemrey) [inline lib/nrf_modem_lib/trace_backends/flash/flash.c:43]: `magic` is one thing, but do we still need the other three variables to be `__noinit` ?

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+{
+	int err, written;
+	size_t to_write = len;
+	const uint8_t *bytes = buf;
+
+	if (!is_initialized) {
+		return -EPERM;
+	}
+
+	while (to_write) {
```
Comment (lemrey) [inline lib/nrf_modem_lib/trace_backends/flash/flash.c:321]: would this loop be a better place to call `trace_processed_callback`?

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+	if (!buf) {
+		return -EINVAL;
+	}
+
+	if (read_offset != 0) {
+		goto read_from_offset;
+	}
+
+	err = fcb_getnext(&trace_fcb, &loc);
+	if (err) {
```
Comment (lemrey) [inline lib/nrf_modem_lib/trace_backends/flash/flash.c:252]: nit: you could reduce the indentation
```
if (err == -ENOTSUP) {
    ....
    return to_read;
}
if (err) {
    return err;
}

read_from_offset:
```

**Review by lemrey — COMMENTED** [review]
Did a first pass

File: lib/nrf_modem_lib/trace_backends/flash/flash.c (hunk 3 above)
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:43]: I'll rewrite this

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+		LOG_ERR("fcb_append_finish failed, err %d", err);
+		return err;
+	}
+
+	stream_buf_written = 0;
+
+	return 0;
+}
+
+static int trace_flash_erase(void)
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:141]: It's already defined in `zephyr/drivers/flash.h`

File: lib/nrf_modem_lib/trace_backends/flash/flash.c (hunk 2 above)
Comment (gregersrygg) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:186]: It looks like the `stream_buf` is preserved on soft-boot, but shouldn't `stream_buf` be written to flash before rebooting? Do we need to preserve `stream_buf`?

Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:186]: I've rewritten this today so that the buffer is flushed when the trace backend is deinitialized.

Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:186]: However, we still need to keep track of how much we have read (which I added to `__noinit` today).

File: lib/nrf_modem_lib/trace_backends/flash/flash.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c]: renamed

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c [hunk 7]
```diff
@@ -294,10 +351,9 @@ static int trace_deinit(void)
 {
 	int err;
 
-	err = trace_backend_deinit();
+	err = trace_backend.deinit();
```
Comment (MirkoCovizzi) [inline lib/nrf_modem_lib/nrf_modem_lib_trace.c:354]: I would add a check to see if the interface function is implemented for every function, even the "default" ones like init, deinit. Backends are potentially implemented by third-parties and they should get an immediate error if they forgot to implement the deinit.

Comment (MirkoCovizzi) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c:354]: Unless it's enforced at compiled time that `init`, `deinit` and `write` are mandatory (but I don't see any compile-time enforcement of this).

**Review by gregersrygg — COMMENTED** [review]
Overall very good improvements for the trace flash backend! I've left a few comments inline.

File: include/modem/nrf_modem_lib_trace.h [hunk 8]
```diff
@@ -51,6 +76,39 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
...
+ *       when the backend is done processing traces.
+ *
+ * @returns Number of bytes stored in the trace backend.
+ */
+size_t nrf_modem_lib_trace_data_size(void);
+
+/**
+ * @brief Read trace data
+ *
+ * @note This operation is only supported with some trace backends.
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply include/modem/nrf_modem_lib_trace.h:92]: Yes, I've added that to `trace_backend.h` but I forgot to add it here. Will add it, thanks!

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c [hunk 9]
```diff
@@ -20,15 +21,22 @@ LOG_MODULE_REGISTER(nrf_modem_lib_trace, CONFIG_NRF_MODEM_LIB_LOG_LEVEL);
 NRF_MODEM_LIB_ON_INIT(trace_init, trace_init_callback, NULL);
 
 K_SEM_DEFINE(trace_sem, 0, 1);
+K_SEM_DEFINE(trace_clear_sem, 0, 1);
 K_SEM_DEFINE(trace_done_sem, 1, 1);
 
+extern struct trace_backend trace_backend;
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c:27]: I will change the struct name to `nrf_modem_lib_trace_backend`.

File: lib/nrf_modem_lib/trace_backends/flash/Kconfig [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+endchoice # NRF_MODEM_TRACE_FLASH_NOSPACE_POLICY
+
+config NRF_MODEM_LIB_TRACE_FLASH_SECTORS
+	int "Number of Flash sectors"
+	default 128
+
+config NRF_MODEM_LIB_TRACE_BACKEND_FLASH_PARTITION_SIZE
+	hex "External flash space reserved for modem traces"
+	default 0x800000
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/Kconfig:50]: ok. I'll change this to half the size.

File: subsys/partition_manager/pm.yml.modem_trace [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 modem_trace:
   placement: {before: [end]}
-  size: CONFIG_PM_PARTITION_SIZE_MODEM_TRACE
+  size: CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_FLASH_PARTITION_SIZE
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply subsys/partition_manager/pm.yml.modem_trace:5]: I don't think it needs to be mentioned as the Kconfig `CONFIG_PM_PARTITION_SIZE_MODEM_TRACE` only existed in a sample.

File: samples/nrf9160/modem_trace_flash/prj.conf [hunk 12]
```diff
@@ -25,11 +30,17 @@ CONFIG_LTE_AUTO_INIT_AND_CONNECT=n
...
 
-# Stream flash
-CONFIG_STREAM_FLASH=y
+# Flash config
+CONFIG_FCB=y
 CONFIG_FLASH=y
 CONFIG_FLASH_MAP=y
 
+# Reduce flash partition size to reduce erase time and flash usage.
+CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_FLASH_PARTITION_SIZE=0x100000
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply samples/nrf9160/modem_trace_flash/prj.conf:39]: I'll move this

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+
+	return to_read;
+
+err:
+	return err;
+}
+
+static int stream_write(const void *buf, size_t len)
+{
+	int err, written;
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:318]: thanks

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 14]
```diff
@@ -102,6 +103,10 @@ Complete the following steps to configure the modem trace level at compile time:
 The application can use the :c:func:`nrf_modem_lib_trace_level_set` function to set the desired trace level.
 Passing ``NRF_MODEM_LIB_TRACE_LEVEL_OFF`` to the :c:func:`nrf_modem_lib_trace_level_set` function disables trace output.
 
+.. note::
+   The current trace level is stored in the modem on ``AT+CFUN=0``.
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:107]: I don't think it is correct to use the `AT+CFUN=0` command to store the trace level. It is more something that occur when CFUN=0 is called.  I'll rewrite this.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 15]
```diff
@@ -521,10 +521,12 @@ Modem libraries
   * Added:
 
     * The :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_LEVEL_OFF` Kconfig option to set the modem trace level to off by default.
+    * The trace flash backend, enabling the application to store trace data to flash and retrieve the data at a later time.
 
   * Updated:
 
     * The minimal value of :kconfig:option:`CONFIG_NRF_MODEM_LIB_SHMEM_RX_SIZE` to meet the requirements of modem firmware 1.3.4.
+    * The :ref:`modem_trace_module` to support backends that store the trace data and retrieve it at a later time.
```
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply doc/nrf/releases/release-notes-changelog.rst:552]: Thanks, this does not necessarily need to be a flash backend, so I'll drop that.

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c (hunk 7 above)
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c:354]: I'll add a check for this in `trace_init`.

**Comment by eivindj-nordic** [comment]
Thank you for your reviews @lemrey, @MirkoCovizzi, @gregersrygg and @divipillai!
I've addressed your changes so please have another look  :)

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (gregersrygg) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst]: What `AT+CFUN=0` does is to persist the current config in the modem, so it will use the same setting after a reboot. The modem will also persist the config on a fixed interval, but I don't remember how many hours between each time.

**Review by gregersrygg — APPROVED** [review]
Very good improvements @eivindj-nordic! Thank you for all small improvements you also added along the way (from offline chat) 🙏

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c [hunk 16]
```diff
@@ -240,12 +298,18 @@ static int trace_init(void)
 {
 	int err;
 
+	if (!trace_backend.init ||
+	    !trace_backend.deinit ||
+	    !trace_backend.write) {
+		LOG_ERR("trace backend does not support init, deinit and write");
```
Comment (MirkoCovizzi) [inline lib/nrf_modem_lib/nrf_modem_lib_trace.c:304]: This code prevents `NULL` dereference, but it can still be improved. I know that clustering these checks looks tidy, but I would separate them. Reading the current log it would appear that if, for example, I forgot to implement deinit, the message will tell me that my backend does not support also init and deinit.
I would also go for "implement" rather than "support" since the interface is implemented.
If you want to keep it clustered, then the log message should say "or" rather than "and", but it still wouldn't be clear for the user which one is not implemented.

Comment (eivindj-nordic) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c:304]: Ok, I will update this to `trace backend must implement init, deinit and write`. It should be easy to check which one is not implemented in the code anyway, so I would rather keep it grouped and small than have three different checks. This is not something that change from backend to backend as the others (read, clear, ...).

Comment (MirkoCovizzi) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c]: That's also ok.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 17]
```diff
@@ -102,9 +103,17 @@ Complete the following steps to configure the modem trace level at compile time:
 The application can use the :c:func:`nrf_modem_lib_trace_level_set` function to set the desired trace level.
 Passing ``NRF_MODEM_LIB_TRACE_LEVEL_OFF`` to the :c:func:`nrf_modem_lib_trace_level_set` function disables trace output.
 
+.. note::
+   If the modem is shut down gracefully before the next boot (by using ``AT+CFUN=0``), it stores the current trace level.
```
Comment (lemrey) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:107]: Saying that `AT+CFUN=0` shuts down the modem is a bit misleading and I would avoid this terminology here. The modem is powered on and some functionality is available after `AT+CFUN=0`. 
I would just say that the trace level is saved upon `AT+CFUN=0`.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 18]
```diff
@@ -102,9 +103,17 @@ Complete the following steps to configure the modem trace level at compile time:
...
+   If the trace level stored in the modem is ``NRF_MODEM_LIB_TRACE_LEVEL_OFF``, to receive ned traces the application must enable traces in the modem using the :c:func:`nrf_modem_lib_trace_level_set` function or by enabling any of the Kconfig options mentioned in step 2.
```
Comment (lemrey) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:108]: `to receive ned traces`?

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c [hunk 19]
```diff
@@ -20,15 +21,22 @@ LOG_MODULE_REGISTER(nrf_modem_lib_trace, CONFIG_NRF_MODEM_LIB_LOG_LEVEL);
 NRF_MODEM_LIB_ON_INIT(trace_init, trace_init_callback, NULL);
 
 K_SEM_DEFINE(trace_sem, 0, 1);
+K_SEM_DEFINE(trace_clear_sem, 0, 1);
 K_SEM_DEFINE(trace_done_sem, 1, 1);
 
+extern struct nrf_modem_lib_trace_backend trace_backend;
+static bool is_space = true;
```
Comment (lemrey) [inline lib/nrf_modem_lib/nrf_modem_lib_trace.c:28]: `has_space`?

File: lib/nrf_modem_lib/trace_backends/flash/Kconfig [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+# Adds Flash to the trace backend choice.
+choice NRF_MODEM_LIB_TRACE_BACKEND
+
+config NRF_MODEM_LIB_TRACE_BACKEND_FLASH
+	bool "Stream modem traces to flash"
+	depends on FCB
```
Comment (lemrey) [inline lib/nrf_modem_lib/trace_backends/flash/Kconfig:12]: Shall we `select` the endmarker kconfig?

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 17 above)
Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:107]: True, this was taken from `nrf91_features.rst`. I will update this.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 18 above)
Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst:108]: removed.

File: doc/nrf/releases/release-notes-changelog.rst [hunk 21]
```diff
@@ -579,11 +579,14 @@ Modem libraries
...
 
     * The :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_LEVEL_OFF` Kconfig option to set the modem trace level to off by default.
+    * The trace flash backend that enables the application to store trace data to flash and retrieve the data later.
 
   * Updated:
 
     * The minimal value of :kconfig:option:`CONFIG_NRF_MODEM_LIB_SHMEM_RX_SIZE` to meet the requirements of modem firmware 1.3.4.
     * The :c:func:`nrf_modem_lib_diag_stats_get` function now returns an error if called when the :ref:`nrf_modem_lib_readme` library has not been initialized.
+    * The trace backend interface to be exposed to the :ref:`modem_trace_module` using the :c:struct:`nrf_modem_lib_trace_backend` struct.
+    * The :ref:`modem_trace_module` to support backends that store the trace data and retrieve the data later.
```
Comment (MirkoCovizzi) [inline doc/nrf/releases/release-notes-changelog.rst:589]:
```suggestion
    * The :ref:`modem_trace_module` to support backends that store the trace data for later retrieval.
```

File: include/modem/nrf_modem_lib_trace.h [hunk 22]
```diff
@@ -51,6 +76,42 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
  */
 int nrf_modem_lib_trace_level_set(enum nrf_modem_lib_trace_level trace_level);
 
+/**
+ * @brief Get the number of bytes stored in the compile-time selected trace backend.
+ *
+ * @note To ensure the retrieved number is correct, this function should only be called
```
Comment (MirkoCovizzi) [inline include/modem/nrf_modem_lib_trace.h:82]:
```suggestion
 * @note To ensure the retrieved number of bytes is correct, this function should only be called
```

File: include/modem/nrf_modem_lib_trace.h [hunk 23]
```diff
@@ -51,6 +76,42 @@ int nrf_modem_lib_trace_processing_done_wait(k_timeout_t timeout);
...
+ * @return read number of bytes, negative errno on failure.
+ */
+int nrf_modem_lib_trace_read(uint8_t *buf, size_t len);
+
+/**
+ * @brief Clear captured trace data
+ *
+ * Clear all captured trace data and prepare for capturing new traces.
+ *
+ * @note This operation is only supported with some trace backends.
```
Comment (MirkoCovizzi) [inline include/modem/nrf_modem_lib_trace.h:109]:
```suggestion
 * @note This operation is implemented only by some trace backends.
```

Same for similar notes.

However, I'd argue we shouldn't add this note to the optional functions. This is the public API and an implementer doesn't really care what other implementations are or aren't implementing.

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c [hunk 24]
```diff
@@ -240,12 +298,18 @@ static int trace_init(void)
 {
 	int err;
 
+	if (!trace_backend.init ||
```
Comment (MirkoCovizzi) [inline lib/nrf_modem_lib/nrf_modem_lib_trace.c:301]: Does this condition fit in one line?

File: lib/nrf_modem_lib/trace_backends/flash/flash.c [hunk 25]
```diff
@@ -0,0 +1,405 @@
...
+
+	err = flash_area_erase(modem_trace_area, 0, TRACE_SIZE);
+	if (err) {
+		LOG_ERR("flash_area_erase error: %d", err);
+	}
+
+	return err;
+}
+
+static const uint32_t magic_val = 0x152ac523;
```
Comment (MirkoCovizzi) [inline lib/nrf_modem_lib/trace_backends/flash/flash.c:152]: Can this be moved to the top together with the others?

File: samples/nrf9160/modem_trace_backend/Kconfig.trace_print_stats [hunk 26]
```diff
@@ -11,9 +11,9 @@ choice NRF_MODEM_LIB_TRACE_BACKEND
 
 config NRF_MODEM_LIB_TRACE_BACKEND_CUSTOM
 	bool "Count modem traces and print to terminal"
-	select CPU_LOAD
-	select FPU
-	select NEWLIB_LIBC_FLOAT_PRINTF
+	depends on CPU_LOAD
```
Comment (MirkoCovizzi) [inline samples/nrf9160/modem_trace_backend/Kconfig.trace_print_stats:14]: Can this change (`select` to `depends on`) be moved to a separate commit since it's separate from the introduction of the flash backend?

File: include/modem/nrf_modem_lib_trace.h (hunk 23 above)
Comment (eivindj-nordic) [reply include/modem/nrf_modem_lib_trace.h:109]: I think it is important for the application engineer to know that there are different backends with different behaviors. There might be cases where someone might want to change the backend during development.

Comment (eivindj-nordic) [reply include/modem/nrf_modem_lib_trace.h:109]: I think supported is ok to use here. If the backend does not support the feature, `nrf_modem_lib_trace.c` returns with `-ENOTSUP`.

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c (hunk 24 above)
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c:301]: yes

File: samples/nrf9160/modem_trace_backend/Kconfig.trace_print_stats (hunk 26 above)
Comment (eivindj-nordic) [reply samples/nrf9160/modem_trace_backend/Kconfig.trace_print_stats:14]: done

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 27]
```diff
@@ -102,9 +103,17 @@ Complete the following steps to configure the modem trace level at compile time:
 The application can use the :c:func:`nrf_modem_lib_trace_level_set` function to set the desired trace level.
 Passing ``NRF_MODEM_LIB_TRACE_LEVEL_OFF`` to the :c:func:`nrf_modem_lib_trace_level_set` function disables trace output.
 
+.. note::
+   The current trace level is stored upon ``AT+CFUN=0``.
```
Comment (divipillai) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:107]:
```suggestion
   The modem stores the current trace level on passing the ``AT+CFUN=0`` command.
```

**Review by divipillai — APPROVED** [review]

File: west.yml [hunk 28]
```diff
@@ -52,7 +52,7 @@ manifest:
     # https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/modules.html
     - name: zephyr
       repo-path: sdk-zephyr
-      revision: 477c1c67430112a57011a2e9202c7b5092b46cff
+      revision: 765155f4716b1f965816e49c38a1b953553be69c
```
Comment (rlubos) [inline west.yml:55]: Please drop this change, manifest is already ahead of this revision.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 29]
```diff
@@ -102,9 +103,17 @@ Complete the following steps to configure the modem trace level at compile time:
 The application can use the :c:func:`nrf_modem_lib_trace_level_set` function to set the desired trace level.
 Passing ``NRF_MODEM_LIB_TRACE_LEVEL_OFF`` to the :c:func:`nrf_modem_lib_trace_level_set` function disables trace output.
 
+.. note::
+   The modem stores the current trace level on passing the ``AT+CFUN=0`` command.
+   If the trace level stored in the modem is ``NRF_MODEM_LIB_TRACE_LEVEL_OFF``, the application must enable traces in the modem using the :c:func:`nrf_modem_lib_trace_level_set` function or by enabling any of the Kconfig options mentioned in step 2.
```
Comment (trantanen) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:108]: Is this the best way to refer to a step (step 2)? It just won't be updated if there are more steps.
@divipillai

File: include/modem/nrf_modem_lib_trace.h (hunk 23 above)
Comment (gregersrygg) [reply include/modem/nrf_modem_lib_trace.h:109]: It would be nice to update the doc-comment that it returns `-ENOTSUP` if it's not supported by the backend. Same for the other similar functions.

Comment (eivindj-nordic) [reply include/modem/nrf_modem_lib_trace.h]: updated

File: doc/nrf/libraries/modem/nrf_modem_lib.rst (hunk 29 above)
Comment (divipillai) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:108]: To refer a step, we normally provide in this manner. For example, https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/app_dev/wifi_coex/index.html#enabling-wi-fi-coexistence.
But in this case we could add this label after line no:96: 
.. _trace_level_options:
and then here, add the following suggestion, so that it is linked to those options:
```suggestion
   If the trace level stored in the modem is ``NRF_MODEM_LIB_TRACE_LEVEL_OFF``, the application must enable traces in the modem using the :c:func:`nrf_modem_lib_trace_level_set` function or by enabling any of the :ref:`aforementioned Kconfig options <trace_level_options>`.
```

Comment (eivindj-nordic) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst]: ok, I've updated the doc with the suggestion.

File: lib/nrf_modem_lib/trace_backends/flash/flash.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (lemrey) [reply lib/nrf_modem_lib/trace_backends/flash/flash.c:140]: ok

File: samples/nrf9160/modem_trace_flash/prj.conf [hunk 30]
```diff
@@ -4,13 +4,17 @@
...
 #
 
+# Allow the application to reboot
+CONFIG_REBOOT=y
+
 # Logging
 CONFIG_LOG=y
-CONFIG_NRF_MODEM_LIB_LOG_LEVEL_DBG=y
-CONFIG_FLASH_LOG_LEVEL_DBG=y
+CONFIG_NRF_MODEM_LIB_LOG_LEVEL_INF=y
```
Comment (lemrey) [inline samples/nrf9160/modem_trace_flash/prj.conf:12]: nit: this is the default

File: samples/nrf9160/modem_trace_flash/prj.conf [hunk 31]
```diff
@@ -25,11 +29,12 @@ CONFIG_LTE_AUTO_INIT_AND_CONNECT=n
...
 CONFIG_NRF_MODEM_LIB_TRACE_BACKEND_FLASH=y
 
-# Stream flash
-CONFIG_STREAM_FLASH=y
+# Flash config
+CONFIG_FCB=y
 CONFIG_FLASH=y
 CONFIG_FLASH_MAP=y
 
+CONFIG_NRF_MODEM_TRACE_FLASH_NOSPACE_SIGNAL=y
```
Comment (lemrey) [inline samples/nrf9160/modem_trace_flash/prj.conf:37]: does it make sense for this sample to choose this option, since it does not show how to handle that event?

Comment (lemrey) [reply samples/nrf9160/modem_trace_flash/prj.conf:37]: if we want to keep that here, we should show how to handle it and how to set up the callback (and when).

**Review by MirkoCovizzi — APPROVED** [review]

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c [hunk 32]
```diff
@@ -136,27 +152,53 @@ static void bps_log(struct k_work *item)
...
+
+void nrf_modem_lib_trace_callback_set(nrf_modem_lib_trace_callback_t callback)
+{
+	if (!callback) {
+		trace_callback = nrf_modem_lib_trace_callback;
+		return;
+	}
+
+	trace_callback = callback;
+}
```
Comment (lemrey) [inline lib/nrf_modem_lib/nrf_modem_lib_trace.c:181]: Trying to remember how this was supposed to work. I think we wanted to handle the scenario where `_NOSPACE_SIGNAL` is selected but the user does not set that callback (or sets it after initializing the library).

The callback has to be set before `nrf_modem_lib_init`, because the trace thread will start automatically when the library is started, and the callback must be ready to receive events by then.

If the user controls when `nrf_modem_lib_init` is called, we have to document `nrf_modem_lib_trace_callback_set` must be called before `nrf_modem_lib_init`. We could enforce this at runtime by checking in `nrf_modem_lib_init` that a callback is set in `nrf_modem_trace.c` and return an error otherwise.

If the user does not control when `nrf_modem_lib_init` is called we would have to ensure a callback exists at the earliest time possible (we can assume `nrf_modem_lib_init` is called during `SYS_INIT` in this case).

I think checking  that `nrf_modem_lib_trace_callback_set` has been called during `nrf_modem_lib_init` gives little possibility to the application to do that in case `nrf_modem_lib_init` is called during `SYS_INIT`, unless the application also setups a `SYS_INIT` function that calls `nrf_modem_lib_trace_callback_set`. That seems like a workaround :)

An alternative that would work in both cases would be to remove `nrf_modem_lib_trace_callback_set` and the weak implementation of `nrf_modem_lib_trace_callback`. Document that if `_NOSPACE_SIGNAL` is selected, the user must implement a function called `nrf_modem_lib_trace_callback` themselves. In the code, we will need to do something like this:

```
#ifdef _NOSPACE_SIGNAL 
extern __weak void nrf_modem_lib_trace_callback(...);
BUILD_ASSERT(nrf_modem_lib_trace_callback, "No nrf_modem_lib_trace_callback() implementation");
#endif

/* when you need to invoke the callback */
#ifdef _NOSPACE_SIGNAL
nrf_modem_lib_trace_callback(event);
#endif
```

File: samples/nrf9160/modem_trace_flash/prj.conf (hunk 31 above)
Comment (trantanen) [reply samples/nrf9160/modem_trace_flash/prj.conf:37]: I think it would be useful to show what should be done (log something in the callback) when `CONFIG_NRF_MODEM_TRACE_FLASH_NOSPACE_SIGNAL` is set because the other option (`CONFIG_NRF_MODEM_TRACE_FLASH_NOSPACE_ERASE_OLDEST`) doesn't require anything from the application.

Comment (eivindj-nordic) [reply samples/nrf9160/modem_trace_flash/prj.conf:37]: I'll change to `ERASE_OLDEST` for now. Might add it later.

File: lib/nrf_modem_lib/nrf_modem_lib_trace.c (hunk 32 above)
Comment (eivindj-nordic) [reply lib/nrf_modem_lib/nrf_modem_lib_trace.c:181]: Discussed offline. Will remove the `_set` function and fail when initializing the flash backend with `_SIGNAL` if no callback is defined. We cannot use BUILD_ASSERT for this as the callback is kept as a symbol until linking. Will add some documentation for this.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 33]
```diff
@@ -102,9 +105,18 @@ Complete the following steps to configure the modem trace level at compile time:
...
+
 During tracing, the integration layer ensures that modem traces are always flushed before the Modem library is re-initialized (including when the modem has crashed).
 The application can synchronize with the flushing of modem traces by calling the :c:func:`nrf_modem_lib_trace_processing_done_wait` function.
 
+For trace backends that support storing of trace data, the application can be notified using the :c:func:`nrf_modem_lib_trace_callback` function if the trace storage becomes full.
+The :c:func:`nrf_modem_lib_trace_callback` must be defined in the application if the :kconfig:option:`CONFIG_NRF_MODEM_TRACE_FLASH_NOSPACE_SIGNAL` Kconfig option is enabled.
+In this case, the application is responsible for reading the trace data with the :c:func:`nrf_modem_lib_trace_read` function if required, before clearing the trace backend storage by calling the :c:func:`nrf_modem_lib_trace_clear` function.
+It is not necessary to turn off modem tracing in this case. However, it is expected that the modem will drop traces when the backend becomes full.
```
Comment (divipillai) [inline doc/nrf/libraries/modem/nrf_modem_lib.rst:118]:
```suggestion
It is not necessary to turn off modem tracing.
However, it is expected that the modem will drop traces when the backend becomes full.
```

**Review by lemrey — APPROVED** [review]

**Review by divipillai — APPROVED** [review]
